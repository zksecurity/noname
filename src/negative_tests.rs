use crate::{
    backends::{
        r1cs::{R1csBn254Field, R1CS},
        Backend,
    },
    circuit_writer::{CircuitWriter, VarInfo},
    compiler::{get_nast, typecheck_next_file_inner, Sources},
    constants::Span,
    error::{ErrorKind, Result},
    imports::FnKind,
    lexer::Token,
    mast::Mast,
    name_resolution::NAST,
    parser::{
        types::{FnSig, GenericParameters},
        ParserCtx,
    },
    type_checker::{FnInfo, FullyQualified, TypeChecker},
    var::Var,
    witness::CompiledCircuit,
};

type R1csBackend = R1CS<R1csBn254Field>;

fn nast_pass(code: &str) -> Result<(NAST<R1csBackend>, usize)> {
    let mut source = Sources::new();
    get_nast(
        None,
        &mut source,
        "example.no".to_string(),
        code.to_string(),
        0,
    )
}

fn tast_pass(code: &str) -> (Result<usize>, TypeChecker<R1csBackend>, Sources) {
    let mut source = Sources::new();
    let mut tast = TypeChecker::<R1csBackend>::new();
    let res = typecheck_next_file_inner(
        &mut tast,
        None,
        &mut source,
        "example.no".to_string(),
        code.to_string(),
        0,
    );

    (res, tast, source)
}

fn mast_pass(code: &str) -> Result<Mast<R1csBackend>> {
    let (_, tast, _) = tast_pass(code);
    crate::mast::monomorphize(tast)
}

fn synthesizer_pass(code: &str) -> Result<CompiledCircuit<R1csBackend>> {
    let mast = mast_pass(code);
    CircuitWriter::generate_circuit(mast?, R1CS::new())
}

#[test]
fn test_return() {
    // no return expected
    let code = r#"
    fn thing(xx: Field) {
        return xx;
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::NoReturnExpected));
}

#[test]
fn test_return_expected() {
    // return expected
    let code = r#"
    fn thing(xx: Field) -> Field {
        let yy = xx + 1;
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::MissingReturn));
}

#[test]
fn test_return_mismatch() {
    // return type mismatch
    let code = r#"
        fn thing(xx: Field) -> Field {
            return true;
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::ReturnTypeMismatch(..)
    ));
}

#[test]
fn test_generic_const_for_loop() {
    let code = r#"
        // generic on const argument
        fn gen(const LEN: Field) -> [Field; LEN] {
            return [0; LEN];
        }

        fn loop() {
            for ii in 0..3 {
                gen(ii);
            }
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::GenericInForLoop));
}

#[test]
fn test_generic_array_for_loop() {
    let code = r#"
        // generic on array argument
        fn gen(arr: [Field; LEN]) -> [Field; LEN] {
            return arr;
        }

        fn loop() {
            let arr = [0; 3];
            for ii in 0..3 {
                gen([0; ii]);
            }
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::GenericInForLoop));
}

#[test]
fn test_generic_missing_parameter_arg() {
    let code = r#"
        fn gen(const len: Field) -> [Field; LEN] {
            return [0; LEN];
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::UndefinedVariable
    ));
}

// #[test]
fn test_generic_symbolic_size_mismatched() {
    let code = r#"
        fn gen(const LEN: Field) -> [Field; 2] {
            return [0; LEN];
        }

        fn main(pub xx: Field) {
            gen(3);
        }
        "#;

    // in theory, this can be caught by the tast phase as it can be checked symbolically.
    // but we can't archive this until
    // - both Array and GenericArray are abstracted into one type with symbolic size.
    // - the `match_expected` and `same_as` functions are replaced by checking rules for different contexts.
    let res = mast_pass(code).err();

    assert!(matches!(
        res.unwrap().kind,
        ErrorKind::ReturnTypeMismatch(..)
    ));
}

#[test]
fn test_const_attr_mismatch() {
    let code = r#"
        struct House {
            room_num: [Field; 2],
        }

        fn House.room(self, const idx: Field) -> Field {
            return self.room_num[idx];
        }

        fn main(pub xx: Field) -> Field {
            let house = House { room_num: [1, 2] };

            // xx is not a constant
            return house.room(xx);
        }
        "#;

    let res = tast_pass(code).0;

    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::ArgumentTypeMismatch(..)
    ));
}

#[test]
fn test_generic_type_mismatched() {
    let code = r#"
        fn gen(const LEN: Field) -> [Field; LEN] {
            return [0; LEN];
        }
        
        fn comp(arr1: [Field; LEN], arr2: [Field; LEN]) {
            for ii in 0..LEN {
                assert_eq(arr1[ii], arr2[ii]);
            }
        }

        fn main(pub xx: Field) {
            let arr1 = gen(2);
            let arr2 = gen(3);
            comp(arr1, arr2);
        }
        "#;

    let res = mast_pass(code).err();
    assert!(matches!(
        res.unwrap().kind,
        ErrorKind::ConflictGenericValue(..)
    ));
}

#[test]
fn test_generic_assignment_type_mismatched() {
    let code = r#"
        fn gen(const LEN: Field) -> [Field; LEN] {
            return [0; LEN];
        }
        
        fn main(pub xx: Field) {
            let mut arr = [0; 3];
            arr = gen(2);
        }
        "#;

    let res = mast_pass(code).err();
    assert!(matches!(
        res.unwrap().kind,
        ErrorKind::AssignmentTypeMismatch(..)
    ));
}

#[test]
fn test_generic_custom_type_mismatched() {
    let code = r#"
        struct Thing {
            xx: [Field; 2],
        }

        fn gen(const LEN: Field) -> [Field; LEN] {
            return [0; LEN];
        }
        
        fn main(pub xx: Field) {
            let arr = gen(3);
            let thing = Thing { xx: arr };
        }
        "#;

    let res = mast_pass(code).err();
    assert!(matches!(
        res.unwrap().kind,
        ErrorKind::InvalidStructFieldType(..)
    ));
}

#[test]
fn test_array_bounds() {
    let code = r#"
        fn gen(const LEN: Field) -> [Field; LEN] {
            return [0; LEN];
        }
        fn main(pub xx: Field) {
            let arr = gen(3);
            arr[3] = 1;
        }
        "#;

    let res = synthesizer_pass(code).err();
    assert!(matches!(
        res.unwrap().kind,
        ErrorKind::ArrayIndexOutOfBounds(..)
    ));
}

#[test]
fn test_iterator_type() {
    let code: &str = r#"
        fn main(pub xx: Field) {
            for ii in xx {
                assert_eq(ii, 3);
            }
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::InvalidIteratorType(..)
    ));
}

#[test]
fn test_iterator_variable_immutable() {
    let code: &str = r#"
        fn main(pub arr: [Field; 3]) {
            for elem in arr {
                elem = 1;
            }
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::AssignmentToImmutableVariable
    ));
}

#[test]
fn test_iterator_variable_redefinition() {
    let code: &str = r#"
        fn main(pub arr: [Field; 3]) {
            let mut var = 5;
            for var in arr {
                assert_eq(var, 1);
            }
        }
        "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::DuplicateDefinition(..)
    ));
}

#[test]
fn test_boolean_and_fail() {
    let code = r#"
    fn thing(xx: Field, yy: Bool) {
        let zz = xx && yy;
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::MismatchType(..)));
}

#[test]
fn test_boolean_or_fail() {
    let code = r#"
    fn thing(xx: Field, yy: Bool) {
        let zz = xx || yy;
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::MismatchType(..)));
}

#[test]
fn test_addition_mismatch() {
    let code = r#"
    fn thing(xx: Field) -> Field {
        let yy = xx + 1;
        return yy + true;
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::MismatchType(..)));
}

#[test]
fn test_multiplication_mismatch() {
    let code = r#"
    fn thing(xx: Field) -> Field {
        let yy = xx * 2;
        return yy + true;
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(res.unwrap_err().kind, ErrorKind::MismatchType(..)));
}

#[test]
fn test_generic_missing_parenthesis() {
    let code = r#"
    fn init_arr(const LEFT: Field) -> [Field; 1 + LEFT * 2] {
        let arr = [0; 1 + (LEFT * 2)];
        return arr;
    }
    "#;

    let res = nast_pass(code).err();
    assert!(matches!(res.unwrap().kind, ErrorKind::MissingParenthesis));
}
fn test_hint_builtin_fn(qualified: &FullyQualified, code: &str) -> Result<usize> {
    let mut source = Sources::new();
    let mut tast = TypeChecker::<R1csBackend>::new();
    // mock a builtin function
    let ctx = &mut ParserCtx::default();
    let mut tokens = Token::parse(0, "calc(val: Field) -> Field;").unwrap();
    let sig = FnSig::parse(ctx, &mut tokens).unwrap();

    fn mocked_builtin_fn<B: Backend>(
        _: &mut CircuitWriter<B>,
        _: &GenericParameters,
        _: &[VarInfo<B::Field, B::Var>],
        _: Span,
    ) -> Result<Option<Var<B::Field, B::Var>>> {
        Ok(None)
    }

    let fn_info = FnInfo {
        kind: FnKind::BuiltIn(sig, mocked_builtin_fn::<R1csBackend>),
        is_hint: true,
        span: Span::default(),
    };

    // add the mocked builtin function
    // note that this should happen in the tast phase, instead of mast phase.
    // currently this function is the only way to mock a builtin function.
    tast.add_monomorphized_fn(qualified.clone(), fn_info);

    typecheck_next_file_inner(
        &mut tast,
        None,
        &mut source,
        "example.no".to_string(),
        code.to_string(),
        0,
    )
}

#[test]
fn test_hint_call_missing_unsafe() {
    let qualified = FullyQualified {
        module: None,
        name: "calc".to_string(),
    };

    let valid_code = r#"
    hint fn calc(val: Field) -> Field;

    fn main(pub xx: Field) {
        let yy = unsafe calc(xx);
    }
    "#;

    let res = test_hint_builtin_fn(&qualified, valid_code);
    assert!(res.is_ok());

    let invalid_code = r#"
    hint fn calc(val: Field) -> Field;

    fn main(pub xx: Field) {
        let yy = calc(xx);
    }
    "#;

    let res = test_hint_builtin_fn(&qualified, invalid_code);
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::ExpectedUnsafeAttribute
    ));
}

#[test]
fn test_nonhint_call_with_unsafe() {
    let code = r#"
    fn calc(val: Field) -> Field {
        return val + 1;
    }

    fn main(pub xx: Field) {
        let yy = unsafe calc(xx);
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::UnexpectedUnsafeAttribute
    ));
}

#[test]
fn test_no_cst_struct_field_prop() {
    let code = r#"
    struct Thing {
        val: Field,
    }

    fn gen(const LEN: Field) -> [Field; LEN] {
        return [0; LEN];
    }

    fn main(pub xx: Field) {
        let thing = Thing { val: xx };

        let arr = gen(thing.val);
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::ArgumentTypeMismatch(..)
    ));
}

#[test]
fn test_mut_cst_struct_field_prop() {
    let code = r#"
    struct Thing {
        val: Field,
    }

    fn gen(const LEN: Field) -> [Field; LEN] {
        return [0; LEN];
    }

    fn main(pub xx: Field) {
        let mut thing = Thing { val: 3 };
        thing.val = xx;

        let arr = gen(thing.val);
        assert_eq(arr[0], xx);
    }
    "#;

    let res = tast_pass(code).0;
    assert!(matches!(
        res.unwrap_err().kind,
        ErrorKind::ArgumentTypeMismatch(..)
    ));
}
