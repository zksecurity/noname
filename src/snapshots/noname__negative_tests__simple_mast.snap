---
source: src/negative_tests.rs
expression: "serde_json::to_string(&mast).unwrap()"
snapshot_kind: text
---
{"functions":[["std/bits/nth_bit",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"nth_bit","span":{"filename_id":0,"start":0,"len":7}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"val","span":{"filename_id":0,"start":8,"len":3}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":13,"len":5}},"attribute":null,"span":{"filename_id":0,"start":8,"len":10}},{"name":{"value":"nth","span":{"filename_id":0,"start":26,"len":3}},"typ":{"kind":{"Field":{"constant":true}},"span":{"filename_id":0,"start":31,"len":5}},"attribute":{"kind":"Const","span":{"filename_id":0,"start":20,"len":5}},"span":{"filename_id":0,"start":20,"len":16}}],"return_type":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":41,"len":5}}},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["std/crypto/poseidon",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"poseidon","span":{"filename_id":0,"start":0,"len":8}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"input","span":{"filename_id":0,"start":9,"len":5}},"typ":{"kind":{"Array":[{"Field":{"constant":false}},2]},"span":{"filename_id":0,"start":16,"len":1}},"attribute":null,"span":{"filename_id":0,"start":9,"len":8}}],"return_type":{"kind":{"Array":[{"Field":{"constant":false}},3]},"span":{"filename_id":0,"start":31,"len":1}}},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["std/builtins/log",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"log","span":{"filename_id":0,"start":0,"len":3}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"var","span":{"filename_id":0,"start":4,"len":3}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":9,"len":5}},"attribute":null,"span":{"filename_id":0,"start":4,"len":10}}],"return_type":null},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["std/bits/check_field_size",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"check_field_size","span":{"filename_id":0,"start":0,"len":16}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"cmp","span":{"filename_id":0,"start":17,"len":3}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":22,"len":5}},"attribute":null,"span":{"filename_id":0,"start":17,"len":10}}],"return_type":null},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["std/int/divmod",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"divmod","span":{"filename_id":0,"start":0,"len":6}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"dividend","span":{"filename_id":0,"start":7,"len":8}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":17,"len":5}},"attribute":null,"span":{"filename_id":0,"start":7,"len":15}},{"name":{"value":"divisor","span":{"filename_id":0,"start":24,"len":7}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":33,"len":5}},"attribute":null,"span":{"filename_id":0,"start":24,"len":14}}],"return_type":{"kind":{"Array":[{"Field":{"constant":false}},2]},"span":{"filename_id":0,"start":43,"len":1}}},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["std/builtins/assert",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"assert","span":{"filename_id":0,"start":0,"len":6}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"condition","span":{"filename_id":0,"start":7,"len":9}},"typ":{"kind":"Bool","span":{"filename_id":0,"start":18,"len":4}},"attribute":null,"span":{"filename_id":0,"start":7,"len":15}}],"return_type":null},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["std/builtins/assert_eq",{"kind":{"BuiltIn":[{"kind":{"Function":"Local"},"name":{"value":"assert_eq","span":{"filename_id":0,"start":0,"len":9}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"lhs","span":{"filename_id":0,"start":10,"len":3}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":15,"len":5}},"attribute":null,"span":{"filename_id":0,"start":10,"len":10}},{"name":{"value":"rhs","span":{"filename_id":0,"start":22,"len":3}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":0,"start":27,"len":5}},"attribute":null,"span":{"filename_id":0,"start":22,"len":10}}],"return_type":null},"native function"]},"is_hint":false,"span":{"filename_id":0,"start":0,"len":0}}],["main",{"kind":{"Native":{"is_hint":false,"sig":{"kind":{"Function":"Local"},"name":{"value":"main","span":{"filename_id":1,"start":53,"len":4}},"generics":{"parameters":{},"resolved_sig":null},"arguments":[{"name":{"value":"xx","span":{"filename_id":1,"start":58,"len":2}},"typ":{"kind":{"Field":{"constant":false}},"span":{"filename_id":1,"start":62,"len":5}},"attribute":null,"span":{"filename_id":1,"start":58,"len":9}}],"return_type":{"kind":{"Custom":{"module":"Local","name":"Thing"}},"span":{"filename_id":1,"start":72,"len":5}}},"body":[{"kind":{"Assign":{"mutable":true,"lhs":{"value":"thing","span":{"filename_id":1,"start":96,"len":5}},"rhs":{"node_id":10,"kind":{"CustomTypeDeclaration":{"custom":{"module":"Local","name":"Thing","span":{"filename_id":1,"start":104,"len":5}},"fields":[[{"value":"val","span":{"filename_id":1,"start":112,"len":3}},{"node_id":9,"kind":{"BigUInt":[3]},"span":{"filename_id":1,"start":117,"len":1}}]]}},"span":{"filename_id":1,"start":104,"len":14}}}},"span":{"filename_id":1,"start":88,"len":30}},{"kind":{"Expr":{"node_id":14,"kind":{"Assignment":{"lhs":{"node_id":12,"kind":{"FieldAccess":{"lhs":{"node_id":11,"kind":{"Variable":{"module":"Local","name":{"value":"thing","span":{"filename_id":1,"start":130,"len":5}}}},"span":{"filename_id":1,"start":130,"len":5}},"rhs":{"value":"val","span":{"filename_id":1,"start":136,"len":3}}}},"span":{"filename_id":1,"start":130,"len":9}},"rhs":{"node_id":13,"kind":{"Variable":{"module":"Local","name":{"value":"xx","span":{"filename_id":1,"start":142,"len":2}}}},"span":{"filename_id":1,"start":142,"len":2}}}},"span":{"filename_id":1,"start":130,"len":14}}},"span":{"filename_id":1,"start":130,"len":14}},{"kind":{"Return":{"node_id":15,"kind":{"Variable":{"module":"Local","name":{"value":"thing","span":{"filename_id":1,"start":161,"len":5}}}},"span":{"filename_id":1,"start":161,"len":5}}},"span":{"filename_id":1,"start":154,"len":13}}],"span":{"filename_id":1,"start":53,"len":114}}},"is_hint":false,"span":{"filename_id":1,"start":53,"len":114}}]],"structs":[["Thing",{"name":"Thing","fields":[["val",{"Field":{"constant":false}}]],"methods":{}}]],"constants":[],"node_types":{"12":{"Field":{"constant":false}},"13":{"Field":{"constant":false}},"8":{"Custom":{"module":"Local","name":"Thing"}},"3":{"Custom":{"module":"Local","name":"Thing"}},"2":{"Field":{"constant":true}},"9":{"Field":{"constant":true}},"11":{"Custom":{"module":"Local","name":"Thing"}},"15":{"Custom":{"module":"Local","name":"Thing"}},"4":{"Custom":{"module":"Local","name":"Thing"}},"10":{"Custom":{"module":"Local","name":"Thing"}},"5":{"Field":{"constant":true}},"6":{"Field":{"constant":false}}},"node_id":15}
